core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java

```diff
@@ -66,34 +64,24 @@ public void setDependencies(RpcManager rpcManager,
 
    @Override
    public CompletableFuture<Void> visitGetAllCommand(InvocationContext ctx, GetAllCommand command) throws Throwable {
-      try {
-         return ctx.shortCircuit(ctx.forkInvocationSync(command));
-      } finally {
-         //when not invoked in an explicit tx's scope the get is non-transactional(mainly for efficiency).
-         //locks need to be released in this situation as they might have been acquired from L1.
-         if (!ctx.isInTxScope()) lockManager.unlockAll(ctx);
-      }
+      if (ctx.isInTxScope())
+         return ctx.continueInvocation();
+
+      return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         lockManager.unlockAll(rCtx);
+         return null;
+      });
    }
 
    @Override
    public CompletableFuture<Void> visitCommitCommand(TxInvocationContext ctx, CommitCommand command) throws Throwable {
-      boolean releaseLocks = releaseLockOnTxCompletion(ctx);
-      try {
-         return ctx.shortCircuit(ctx.forkInvocationSync(command));
-      } catch (OutdatedTopologyException e) {
-         releaseLocks = false;
-         throw e;
-      } finally {
-         if (releaseLocks) lockManager.unlockAll(ctx);
-      }
-   }
+      return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         if (throwable instanceof OutdatedTopologyException)
+            throw throwable;
 
-   protected final Object invokeNextAndCommitIf1Pc(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
-      Object result = ctx.forkInvocationSync(command);
-      if (command.isOnePhaseCommit() && releaseLockOnTxCompletion(ctx)) {
-         lockManager.unlockAll(ctx);
-      }
-      return result;
+         releaseLockOnTxCompletion(((TxInvocationContext) rCtx));
```

Comment:
Yep, avoiding capturing locals in the lambdas is both hard to to consistently and uglier.